patch taken from: https://sourceforge.net/p/p7zip/bugs/230/#c36e

       The getpassphrase() function is identical to getpass(), except that it
       reads and returns a string of up to 257 characters in length.
...
       Upon successful completion, getpass() returns a pointer to a null-
       terminated string of at most 9 bytes that were read from the terminal
       device. If an error is encountered, the terminal state is restored and
       a null pointer is returned.

diff -wpruN '--exclude=*.orig' a~/CPP/7zip/UI/Console/UserInputUtils.cpp a/CPP/7zip/UI/Console/UserInputUtils.cpp
--- a~/CPP/7zip/UI/Console/UserInputUtils.cpp	1970-01-01 00:00:00
+++ a/CPP/7zip/UI/Console/UserInputUtils.cpp	1970-01-01 00:00:00
@@ -89,12 +89,21 @@ UString GetPassword(CStdOutStream *outSt
     outStream->Flush();
   }
 #ifdef ENV_HAVE_GETPASS
+#if defined(__illumos__)
+  AString oemPassword = getpassphrase("");
+#else
   AString oemPassword = getpass("");
+#endif
+
   if ( (verify) && (outStream) )
   {
     (*outStream) << "Verify password (will not be echoed) :";
     outStream->Flush();
+#if defined(__illumos__)
+    AString oemPassword2 = getpassphrase("");
+#else
     AString oemPassword2 = getpass("");
+#endif
     if (oemPassword != oemPassword2) throw "password verification failed";
   }
   return MultiByteToUnicodeString(oemPassword, CP_OEMCP);
